File: crates/rullama-rag/examples/code_analysis.rs
//! Code Analysis — Definitions, References, and Call Graphs
//!
//! Demonstrates how to use the code analysis module to extract symbol
//! definitions, find references, and build call graph structures from
//! source code using AST-based (tree-sitter) parsing.
//!
//! Run:
//! ```sh
//! cargo run -p rullama-knowledge --example code_analysis --features code-analysis,rag
//! ```

use std::collections::HashMap;

use rullama_rag::code_analysis::types::{
    CallEdge, CallGraphNode, DefinitionResult, PrecisionLevel, SymbolId,
};
use rullama_rag::code_analysis::{
    Definition, HybridRelationsProvider, RelationsProvider, SymbolKind,
};
use rullama_rag::rag::indexer::FileInfo;

fn main() -> anyhow::Result<()> {
    println!("=== Brainwires Code Analysis Example ===\n");

    // ── Step 1: Create the hybrid relations provider ─────────────────
    println!("--- Step 1: Initialize Relations Provider ---\n");

    let provider = HybridRelationsProvider::new()?;

    // Check language support and precision levels
    let languages = ["Rust", "Python", "JavaScript", "TypeScript", "Go", "Java"];
    println!("{:<14} {:<10} Precision", "Language", "Supported");
    println!("{:-<44}", "");
    for lang in &languages {
        let supported = provider.supports_language(lang);
        let precision = provider.precision_level(lang);
        println!(
            "{:<14} {:<10} {}",
            lang,
            if supported { "yes" } else { "no" },
            precision.description(),
        );
    }
    println!();

    // ── Step 2: Extract definitions from a Rust source snippet ───────
    println!("--- Step 2: Extract Definitions ---\n");

    let rust_source = r#"
use std::collections::HashMap;

/// Configuration for the application.
pub struct Config {
    pub host: String,
    pub port: u16,
    pub debug: bool,
}

impl Config {
    /// Create a new Config with defaults.
    pub fn new() -> Self {
        Self {
            host: "localhost".to_string(),
            port: 8080,
            debug: false,
        }
    }

    /// Validate the configuration.
    pub fn validate(&self) -> Result<(), String> {
        if self.port == 0 {
            return Err("port must be non-zero".into());
        }
        Ok(())
    }
}

/// Process an incoming request using the given config.
pub fn process_request(config: &Config, data: &str) -> String {
    format!("Processed on {}:{}: {}", config.host, config.port, data)
}

/// Start the main server loop.
fn main() {
    let config = Config::new();
    config.validate().expect("invalid config");
    let result = process_request(&config, "hello world");
    println!("{}", result);
}
"#;

    let file_info = FileInfo {
        path: std::path::PathBuf::from("/demo/src/main.rs"),
        relative_path: "src/main.rs".to_string(),
        root_path: "/demo".to_string(),
        project: Some("demo".to_string()),
        extension: Some("rs".to_string()),
        language: Some("Rust".to_string()),
        content: rust_source.to_string(),
        hash: "demo-hash-001".to_string(),
    };

    let definitions = provider.extract_definitions(&file_info)?;

    println!("Found {} definitions in src/main.rs:\n", definitions.len());
    println!(
        "{:<6} {:<22} {:<12} {:<10} {:<10} Signature",
        "Line", "Name", "Kind", "Visibility", "Has Docs"
    );
    println!("{:-<90}", "");

    for def in &definitions {
        let sig_preview = if def.signature.len() > 30 {
            format!("{}...", &def.signature[..30])
        } else {
            def.signature.clone()
        };

        println!(
            "{:<6} {:<22} {:<12} {:<10} {:<10} {}",
            def.symbol_id.start_line,
            def.symbol_id.name,
            def.symbol_id.kind.display_name(),
            format!("{:?}", def.visibility),
            if def.doc_comment.is_some() {
                "yes"
            } else {
                "no"
            },
            sig_preview
        );
    }
    println!();

    // ── Step 3: Build a symbol index and find references ─────────────
    println!("--- Step 3: Find References ---\n");

    // Build a symbol index from the definitions we extracted
    let mut symbol_index: HashMap<String, Vec<Definition>> = HashMap::new();
    for def in &definitions {
        symbol_index
            .entry(def.symbol_id.name.clone())
            .or_default()
            .push(def.clone());
    }

    let references = provider.extract_references(&file_info, &symbol_index)?;

    println!("Found {} references:\n", references.len());
    println!("{:<6} {:<10} {:<16} Target", "Line", "Col", "Kind");
    println!("{:-<60}", "");

    for reference in &references {
        println!(
            "{:<6} {:<10} {:<16} {}",
            reference.start_line,
            reference.start_col,
            format!("{:?}", reference.reference_kind),
            reference.target_symbol_id,
        );
    }
    println!();

    // ── Step 4: Demonstrate SymbolId and DefinitionResult types ──────
    println!("--- Step 4: Symbol Identification ---\n");

    // Create symbol IDs for cross-referencing
    let sym_config = SymbolId::new("src/main.rs", "Config", SymbolKind::Struct, 5, 0);
    let sym_new = SymbolId::new("src/main.rs", "new", SymbolKind::Method, 13, 4);
    let sym_process = SymbolId::new(
        "src/main.rs",
        "process_request",
        SymbolKind::Function,
        33,
        0,
    );

    println!("Symbol IDs (for storage/lookup):");
    println!("  Config:          {}", sym_config.to_storage_id());
    println!("  Config::new:     {}", sym_new.to_storage_id());
    println!("  process_request: {}", sym_process.to_storage_id());
    println!();

    // Convert a Definition to a DefinitionResult (the MCP-facing type)
    if let Some(first_def) = definitions.first() {
        let result = DefinitionResult::from(first_def);
        println!("DefinitionResult for '{}':", result.name);
        println!("  File:      {}", result.file_path);
        println!("  Kind:      {:?}", result.kind);
        println!("  Lines:     {}-{}", result.start_line, result.end_line);
        println!("  Signature: {}", result.signature);
        if let Some(ref doc) = result.doc_comment {
            println!("  Docs:      {}", doc);
        }
    }
    println!();

    // ── Step 5: Demonstrate CallEdge and CallGraphNode types ─────────
    println!("--- Step 5: Call Graph Structures ---\n");

    // Build a small call graph manually to demonstrate the data model
    let call_edges = vec![
        CallEdge {
            caller_id: "src/main.rs:main:39:0".to_string(),
            callee_id: "src/main.rs:new:13:4".to_string(),
            call_site_file: "src/main.rs".to_string(),
            call_site_line: 40,
            call_site_col: 18,
        },
        CallEdge {
            caller_id: "src/main.rs:main:39:0".to_string(),
            callee_id: "src/main.rs:validate:23:4".to_string(),
            call_site_file: "src/main.rs".to_string(),
            call_site_line: 41,
            call_site_col: 11,
        },
        CallEdge {
            caller_id: "src/main.rs:main:39:0".to_string(),
            callee_id: "src/main.rs:process_request:33:0".to_string(),
            call_site_file: "src/main.rs".to_string(),
            call_site_line: 42,
            call_site_col: 18,
        },
    ];

    println!("Call edges from main():");
    for edge in &call_edges {
        println!(
            "  main (line {}) → {} at {}:{}",
            edge.call_site_line,
            edge.callee_id.split(':').nth(1).unwrap_or("?"),
            edge.call_site_file,
            edge.call_site_line
        );
    }
    println!();

    // Build a CallGraphNode tree
    let call_graph = CallGraphNode {
        name: "main".to_string(),
        kind: SymbolKind::Function,
        file_path: "src/main.rs".to_string(),
        line: 39,
        children: vec![
            CallGraphNode {
                name: "Config::new".to_string(),
                kind: SymbolKind::Method,
                file_path: "src/main.rs".to_string(),
                line: 13,
                children: vec![],
            },
            CallGraphNode {
                name: "Config::validate".to_string(),
                kind: SymbolKind::Method,
                file_path: "src/main.rs".to_string(),
                line: 23,
                children: vec![],
            },
            CallGraphNode {
                name: "process_request".to_string(),
                kind: SymbolKind::Function,
                file_path: "src/main.rs".to_string(),
                line: 33,
                children: vec![],
            },
        ],
    };

    println!("Call graph tree:");
    print_call_tree(&call_graph, 0);
    println!();

    // ── Step 6: Precision levels and configuration ───────────────────
    println!("--- Step 6: Configuration ---\n");

    let config = rullama_rag::code_analysis::RelationsConfig::default();
    println!("Default RelationsConfig:");
    println!("  enabled:        {}", config.enabled);
    println!("  max_call_depth: {}", config.max_call_depth);
    println!();

    println!("Precision levels:");
    for level in [PrecisionLevel::Medium, PrecisionLevel::Low] {
        println!("  {:?} — {}", level, level.description());
    }

    println!("\n=== Done ===");
    Ok(())
}

/// Recursively print a call graph tree with indentation.
fn print_call_tree(node: &CallGraphNode, depth: usize) {
    let indent = "  ".repeat(depth);
    println!(
        "{}{} ({}) — {}:{}",
        indent,
        node.name,
        node.kind.display_name(),
        node.file_path,
        node.line
    );
    for child in &node.children {
        print_call_tree(child, depth + 1);
    }
}

File: deno/examples/knowledge/code_analysis.ts
// Example: Code Analysis
// Demonstrates symbol extraction, reference finding, call graph construction, and repo map formatting.
// Run: deno run deno/examples/cognition/code_analysis.ts

import {
  buildCallGraph,
  CallGraph,
  createSymbolId,
  definitionToStorageId,
  findReferences,
  RepoMap,
  symbolIdToStorageId,
  symbolKindDisplayName,
} from "@rullama/knowledge";

import type {
  CallEdge,
  CallGraphNode,
  CodeAnalysisDefinition,
  SymbolKind,
} from "@rullama/knowledge";

async function main() {
  console.log("=== rullama Code Analysis Example ===\n");

  // 1. Check supported languages
  console.log("--- Step 1: Language Support ---\n");

  const extensions = RepoMap.supportedExtensions();
  console.log(`Supported extensions: ${extensions.join(", ")}`);
  console.log();

  const testExts = ["ts", "js", "py", "rs", "go", "java"];
  console.log(`${"Extension".padEnd(12)} ${"Supported".padEnd(12)} Language`);
  console.log("-".repeat(44));
  for (const ext of testExts) {
    const supported = RepoMap.supportsExtension(ext);
    const language = RepoMap.languageForExtension(ext) ?? "(none)";
    console.log(
      `${ext.padEnd(12)} ${(supported ? "yes" : "no").padEnd(12)} ${language}`,
    );
  }
  console.log();

  // 2. Extract definitions from a TypeScript source snippet
  console.log("--- Step 2: Extract Definitions ---\n");

  const tsSource = `
import { EventEmitter } from "events";

/** Configuration for the application. */
export interface Config {
  host: string;
  port: number;
  debug: boolean;
}

/** Create a new Config with defaults. */
export function createConfig(): Config {
  return {
    host: "localhost",
    port: 8080,
    debug: false,
  };
}

/** Validate the configuration. */
export function validateConfig(config: Config): boolean {
  if (config.port === 0) {
    throw new Error("port must be non-zero");
  }
  return true;
}

/** Process an incoming request using the given config. */
export function processRequest(config: Config, data: string): string {
  return \`Processed on \${config.host}:\${config.port}: \${data}\`;
}

/** Start the main server loop. */
function main() {
  const config = createConfig();
  validateConfig(config);
  const result = processRequest(config, "hello world");
  console.log(result);
}
`;

  const definitions = RepoMap.extractSymbols({
    filePath: "src/server.ts",
    content: tsSource,
    rootPath: "/demo",
    project: "demo",
  });

  console.log(`Found ${definitions.length} definitions in src/server.ts:\n`);
  console.log(
    `${"Line".padEnd(6)} ${"Name".padEnd(22)} ${"Kind".padEnd(12)} ${
      "Visibility".padEnd(12)
    } Signature`,
  );
  console.log("-".repeat(90));

  for (const def of definitions) {
    const sigPreview = def.signature.length > 40
      ? def.signature.slice(0, 40) + "..."
      : def.signature;

    console.log(
      `${String(def.symbolId.startLine).padEnd(6)} ` +
        `${def.symbolId.name.padEnd(22)} ` +
        `${symbolKindDisplayName(def.symbolId.kind).padEnd(12)} ` +
        `${def.visibility.padEnd(12)} ` +
        `${sigPreview}`,
    );
  }
  console.log();

  // 3. Build a symbol index and find references
  console.log("--- Step 3: Find References ---\n");

  const symbolIndex = new Map<string, CodeAnalysisDefinition[]>();
  for (const def of definitions) {
    const existing = symbolIndex.get(def.symbolId.name) ?? [];
    existing.push(def);
    symbolIndex.set(def.symbolId.name, existing);
  }

  const references = findReferences(
    "src/server.ts",
    tsSource,
    symbolIndex,
    "/demo",
    "demo",
  );

  console.log(`Found ${references.length} references:\n`);
  console.log(
    `${"Line".padEnd(6)} ${"Col".padEnd(8)} ${"Kind".padEnd(18)} Target`,
  );
  console.log("-".repeat(60));

  for (const ref of references) {
    console.log(
      `${String(ref.startLine).padEnd(6)} ` +
        `${String(ref.startCol).padEnd(8)} ` +
        `${ref.referenceKind.padEnd(18)} ` +
        `${ref.targetSymbolId}`,
    );
  }
  console.log();

  // 4. Symbol identification
  console.log("--- Step 4: Symbol Identification ---\n");

  const symConfig = createSymbolId(
    "src/server.ts",
    "Config",
    "interface",
    5,
    0,
  );
  const symCreate = createSymbolId(
    "src/server.ts",
    "createConfig",
    "function",
    12,
    0,
  );
  const symProcess = createSymbolId(
    "src/server.ts",
    "processRequest",
    "function",
    30,
    0,
  );

  console.log("Symbol IDs (for storage/lookup):");
  console.log(`  Config:          ${symbolIdToStorageId(symConfig)}`);
  console.log(`  createConfig:    ${symbolIdToStorageId(symCreate)}`);
  console.log(`  processRequest:  ${symbolIdToStorageId(symProcess)}`);
  console.log();

  if (definitions.length > 0) {
    const first = definitions[0];
    const storageId = definitionToStorageId(first);
    console.log(`First definition storage ID: ${storageId}`);
    console.log(`  Name:      ${first.symbolId.name}`);
    console.log(`  File:      ${first.symbolId.filePath}`);
    console.log(`  Kind:      ${symbolKindDisplayName(first.symbolId.kind)}`);
    console.log(`  Lines:     ${first.symbolId.startLine}-${first.endLine}`);
    console.log(`  Signature: ${first.signature}`);
    if (first.docComment) {
      console.log(`  Docs:      ${first.docComment}`);
    }
  }
  console.log();

  // 5. Build a call graph
  console.log("--- Step 5: Call Graph ---\n");

  const files = new Map<string, string>();
  files.set("src/server.ts", tsSource);

  const callGraph = buildCallGraph(definitions, files);
  console.log(
    `Call graph: ${callGraph.nodes.size} nodes, ${callGraph.edges.length} edges`,
  );
  console.log();

  console.log("Call edges:");
  for (const edge of callGraph.edges) {
    const callerName = edge.callerId.split(":")[1] ?? edge.callerId;
    const calleeName = edge.calleeId.split(":")[1] ?? edge.calleeId;
    console.log(
      `  ${callerName} (line ${edge.callSiteLine}) -> ${calleeName}`,
    );
  }
  console.log();

  // Display call tree from a root node
  for (const [nodeId] of callGraph.nodes) {
    const tree = callGraph.calleeTree(nodeId, 3);
    if (tree && tree.children.length > 0) {
      console.log(`Call tree from ${tree.name}:`);
      printCallTree(tree, 0);
      console.log();
    }
  }

  // 6. Repo map formatting
  console.log("--- Step 6: Repo Map ---\n");

  const repoMap = RepoMap.formatRepoMap(definitions);
  console.log(repoMap);

  console.log("\n=== Done ===");
}

/** Recursively print a call graph tree with indentation. */
function printCallTree(node: CallGraphNode, depth: number): void {
  const indent = "  ".repeat(depth);
  console.log(
    `${indent}${node.name} (${
      symbolKindDisplayName(node.kind)
    }) -- ${node.filePath}:${node.line}`,
  );
  for (const child of node.children) {
    printCallTree(child, depth + 1);
  }
}

await main();
